### PR TITLE
#167484614 Highlight and comment on an article

### PR DESCRIPTION
**Pull request by malaba6**
### What does this PR do?
- Adds the functionality to highlight a text of an article and makes it possible to comment on it

### Description of Task to be completed?
- Add Highlight model in association with the user and articles models
- Add validation for text highlighting
- Add user controller to create a text Highlight

### How should this be manually tested?
In Postman or Swagger or any other tool of your choice visit these endpoints:
*POST*`/api/v1/articles/:slug/highlights` To add a text highlight

### What are the relevant pivotal tracker stories?
[Users should be able to highlight and comment on any text in an article](https://www.pivotaltracker.com/story/show/#167484614)

### Screenshots? 
*In Postman*
Screen Shot 2019-08-30 at 13 30 52
*In Swagger*
Screen Shot 2019-08-30 at 13 29 31

### CODE REVIEW

File: src/sequelize/models/User.js [hunk 1]
```diff
@@ -80,6 +80,7 @@ export default (sequelize, DataTypes) => {
     User.hasMany(models.Reaction, { foreignKey: 'userId', as: 'LikerOrDisliker', onDelete: 'CASCADE' });
     User.hasMany(models.Comment, { foreignKey: 'userId', as: 'Commenter', onDelete: 'CASCADE' });
     User.hasMany(models.Bookmark, { foreignKey: 'userId', onDelete: 'CASCADE' });
+    User.hasMany(models.Highlight, { foreignKey: 'userId', onDelete: 'CASCADE' });
```
Comment (harerakalex) [inline src/sequelize/models/User.js:83]: I wonder why you didn't give alias name this association? But otherwise looks good.

File: src/index.js [hunk 2]
```diff
@@ -45,12 +45,12 @@ app.use((req, res) => {
   res.status(404).json({ error: 'route not found' });
 });
 
-app.use((error, req, res, next) => {
-  res.status(500).json({
-    error: error.message,
-    next
-  });
-});
+// app.use((error, req, res, next) => {
```
Comment (Jkadhuwa) [inline src/index.js]: Why did you comment these lines out?

**Review by placideirandora — CHANGES_REQUESTED** [review]
You should fix this issue of an invalid index. The word does exist but the index doesn't.
Screen Shot 2019-09-02 at 15 59 15
Screen Shot 2019-09-02 at 16 00 23

**Comment by malaba6** [comment]
> You should fix this issue of an invalid index. The word does exist but the index doesn't.
> Screen Shot 2019-09-02 at 15 59 15
> Screen Shot 2019-09-02 at 16 00 23

Right away! Fixed....

File: src/controllers/higlightController.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+      if (!article) {
+        return res.status(404).json({
+          error: `Article ${articleSlug} does not exist`
+        });
+      }
+
+      const { body } = article;
+      let highlightedText = 'highlighted';
+      let { comment } = payload;
+      const { startIndex, text } = payload;
```
Comment (mkiterian) [inline src/controllers/higlightController.js]: why not send a startIndex and a stopIndex?

File: src/controllers/higlightController.js [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      if (startIndex > body.length) {
+        return res.status(404).json({
+          error: 'Invalid index'
+        });
+      }
+      const textToHighlight = body.slice(startIndex);
+
+      if (!textToHighlight.includes(text)) {
```
Comment (mkiterian) [inline src/controllers/higlightController.js]: I recommend you send a startIndex and a stopIndex which would alleviate the need for this check, you'd only check that the stop index does not exceed the body length

**Review by mkiterian — CHANGES_REQUESTED** [review]

File: src/controllers/higlightController.js (hunk 3 above)
Comment (malaba6) [reply src/controllers/higlightController.js]: Changed!

File: src/controllers/higlightController.js (hunk 4 above)
Comment (malaba6) [reply src/controllers/higlightController.js]: Ok! makes sense!

**Pull request merged**